### PR TITLE
chore: fix master and v3 branches

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes https://jira.d2iq.com/browse/COPS-7145

https://github.com/mesosphere/traefik-forward-auth/releases/tag/v3.0.0 was released with https://github.com/mesosphere/traefik-forward-auth/pull/41 feature. But the `v3` branch was cut before that work went in. Then there was a big refactor that was done in https://github.com/mesosphere/traefik-forward-auth/commit/7e8ab37 and merged to `v3` branch but the [PR to master](https://github.com/mesosphere/traefik-forward-auth/pull/50) was never merged in. Then there were some more work that went into `v3` and most of it did not make to `master`.

This PR is to get `master` back to expected state that includes both [Jared's clusterstorage feature](https://github.com/mesosphere/traefik-forward-auth/pull/41) and the changes from the `v3` branch. After this is merged we can create a new `release/v3` branch from this tip that can then be used for `v3` development.

---

Jason and I cherry-picked each commit from `v3`. The first 4 had the most merge conflicts.
When reviewing pay special attention to 39539aa1f682010131e5d12413cae2072a6084d8 as that one made some additional changes from the cherry-pick commit. The [original commit](https://github.com/mesosphere/traefik-forward-auth/commit/36c3eee4c9fa262064848d4ddaca6652b96763b5#diff-2b0bf79e19fe4eb1540beb0b6d88642e3c87ab38de6103290a4e4f775e377173R272) added a `sessionCookie` which is no longer needed with `userinfo` introduced in the [clusterstorage PR](https://github.com/mesosphere/traefik-forward-auth/pull/41).

image

### CODE REVIEW

**Comment by dkoshkin** [comment]
Shoutout to @jkoelker for the work here 👏 !

File: internal/authorization/rbac/rbac.go [hunk 1]
```diff
@@ -153,7 +155,7 @@ func verbMatches(rule *rbacv1.PolicyRule, requestedVerb string) bool {
 		if ruleVerb == rbacv1.VerbAll {
 			return true
 		}
-		if strings.ToLower(ruleVerb) == strings.ToLower(requestedVerb) {
+		if strings.EqualFold(ruleVerb, requestedVerb) {
```
Comment (joejulian) [inline internal/authorization/rbac/rbac.go:170]: Is there any security ramifications of using EqualFold? This does more than just case-insensitive comparisons? cc @mhrabovcin

File: internal/authorization/rbac/rbac.go [hunk 2]
```diff
@@ -173,3 +175,13 @@ func nonResourceURLMatches(rule *rbacv1.PolicyRule, requestedURL string) bool {
...
 }
+
+// compareSubjects determines whether subjects names are equal (string equality is used).
+// If caseInsensitive is true, the case of the characters is ignored, meaning "UserName"
+// would be considered equal to "username".
+func compareSubjects(s1, s2 string, caseInsensitive bool) bool {
+	if caseInsensitive == false {
+		return s1 == s2
+	}
+	return strings.EqualFold(s1, s2)
```
Comment (joejulian) [inline internal/authorization/rbac/rbac.go:216]: Is there any security ramifications of using EqualFold? This does more than just case-insensitive comparisons? cc @mhrabovcin

**Review by mhrabovcin — APPROVED** [review]

**Comment by jkoelker** [comment]
LGTM, I'll refrain from approving, since I helped with the code side, but I went through the remaining commits and it looks good to me.

**Comment by dkoshkin** [comment]
Using https://github.com/mesosphere/kubernetes-base-addons/tree/dkoshkin-tfa-clusterstorage

Tested manually by setting custom values on the addon:
```
        - name: traefik-forward-auth
          enabled: true
          values: |
            clusterStorage:
              enabled: true
              namespace: kubeaddons
```

Verified the login still worked and saw this:
```
➜  dkkonvoy-tfa-cops k view-secret -n kubeaddons tfa-claims-nlmrm
Choosing key: userInfo
{"username":"hungry_banach","email":"hungry_banach","groups":[]}
```

**Review by joejulian — APPROVED** [review]

**Pull request merged**